cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -1030,6 +1031,32 @@ class AsofJoinNode : public ExecNode {
     return match.indices()[0];
   }
 
+  static Result<std::vector<col_index_t>> GetIndicesOfOnKey(
+      const std::vector<std::shared_ptr<Schema>>& input_schema, const FieldRef& on_key) {
+    size_t n_input = input_schema.size();
+    std::vector<col_index_t> indices_of_on_key(n_input);
+    for (size_t i = 0; i < n_input; ++i) {
+      ARROW_ASSIGN_OR_RAISE(indices_of_on_key[i],
+                            FindColIndex(*input_schema[i], on_key, "on"));
```

Comment:
I thought substrait/FieldRef is index based already - why do we need to find index here?

Reply:
The proto uses `Expression` to allow for various ways to specify the key; it can be thought of as future-proof. We can choose which of these ways to support in Arrow Substrait in a first version. The code here supports any kind of `FieldRef`, be it named or indexed, that Arrow Substrait may set up. The Arrow Substrait code only supports an indexed `FieldRef` (see [here](https://github.com/apache/arrow/blob/e8d54ea765ae7ba63b8f42c29ec855d656e85dc8/cpp/src/arrow/engine/substrait/expression_internal.cc#L150)), but it can be extended.

Reply:
I am not sure we need this function - in practice this substrait field ref should always be index based..

Reply:
Note that the code here already appeared in the pre-PR code, just moved a bit.

I suspect going down the path you describe, which I'm willing to try, would be more complicated than you suggest because even `AsofJoinNodeOptions` does not currently support distinct field-references per input (be them index- or name- based). Therefore, I think the required fix would involve many places in the code (and we would normally have allocated a separate task for such a fix).